source/extensions/filters/network/ratelimit/ratelimit.cc

```diff
@@ -18,16 +23,44 @@ namespace RateLimitFilter {
 Config::Config(const envoy::extensions::filters::network::ratelimit::v3::RateLimit& config,
                Stats::Scope& scope, Runtime::Loader& runtime)
     : domain_(config.domain()), stats_(generateStats(config.stat_prefix(), scope)),
-      runtime_(runtime), failure_mode_deny_(config.failure_mode_deny()) {
+      runtime_(runtime), failure_mode_deny_(config.failure_mode_deny()),
+      request_headers_(Http::RequestHeaderMapImpl::create()),
+      response_headers_(Http::ResponseHeaderMapImpl::create()),
+      response_trailers_(Http::ResponseTrailerMapImpl::create()) {
   for (const auto& descriptor : config.descriptors()) {
     RateLimit::Descriptor new_descriptor;
     for (const auto& entry : descriptor.entries()) {
       new_descriptor.entries_.push_back({entry.key(), entry.value()});
+      substitution_formatters_.push_back(
+          std::make_shared<Formatter::FormatterImpl>(entry.value(), false));
     }
     descriptors_.push_back(new_descriptor);
   }
 }
 
+void Config::applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                        StreamInfo::StreamInfo& stream_info) {
+
+  std::vector<RateLimit::Descriptor> dynamic_descriptors = std::vector<RateLimit::Descriptor>();
```

Comment:
I don't quite understand what this function is doing. I think all you need to do is run through the formatters and apply all of them.

Reply:
It's applying each unique Formatter stored in the vector created during config in matched order as the descriptor values are stored once the stream_info is available.

Reply:
See my above comment. You should only need to store the formatters and then run through them.

/wait

Reply:
The limit function requires a descriptor as input. That would have to change for this change to go through. Do we want to do that.